File: content-app/generate_data.js
import { promises as fs } from 'fs';
import path from 'path';

async function getMdFiles(directory) {
    const mdFiles = [];
    async function walk(dir) {
        const files = await fs.readdir(dir);
        for (const file of files) {
            const filePath = path.join(dir, file);
            const stat = await fs.stat(filePath);
            if (stat.isDirectory()) {
                await walk(filePath);
            } else if (file.endsWith('.md')) {
                mdFiles.push(filePath);
            }
        }
    }
    await walk(directory);
    return mdFiles;
}

async function extractMetadata(filePath) {
    const content = await fs.readFile(filePath, 'utf-8');
    const lines = content.split('\n');
    let title = '';
    let date = '';
    for (const line of lines) {
        if (line.startsWith('# ')) {
            title = line.replace('# ', '').trim();
        } else if (line.startsWith('Date: ')) {
            date = line.replace('Date: ', '').trim();
        }
        if (title && date) break;
    }
    if (!title || !date) {
        title = path.basename(filePath, '.md');
        const stat = await fs.stat(filePath);
        date = new Date(stat.mtime).toISOString().split('T')[0];
    }
    return { title, date };
}

async function generateDataTs(mdFiles, filepath) {
    const data = await Promise.all(mdFiles.map(extractMetadata));
    data.sort((a, b) => new Date(b.date) - new Date(a.date));

    const tsContent = `export const data = ${JSON.stringify(data, null, 2)};`;

    await fs.writeFile(filepath, tsContent, 'utf-8');
}

async function main() {
    const blogsDir = 'pages/blogs';
    const mdFiles = await getMdFiles(blogsDir);
    await generateDataTs(mdFiles, 'data/blogs.ts');

    const weeklyDir = 'pages/weekly';
    const weeklyMdFiles = await getMdFiles(weeklyDir);
    await generateDataTs(weeklyMdFiles, 'data/weekly.ts');

    console.log("data.ts 已生成");
}

main().catch(console.error);

File: content-app/generate_data.py
import os
import json
from datetime import datetime

def get_md_files(directory):
    md_files = []
    for root, dirs, files in os.walk(directory):
        for file in files:
            if file.endswith('.md'):
                md_files.append(os.path.join(root, file))
    return md_files

def extract_metadata(file_path):
    with open(file_path, 'r', encoding='utf-8') as file:
        content = file.read()
        lines = content.split('\n')
        title = ''
        date = ''
        for line in lines:
            if line.startswith('# '):
                title = line.replace('# ', '').strip()
            elif line.startswith('Date: '):
                date = line.replace('Date: ', '').strip()
            if title and date:
                break
        if not title or not date:
            title = os.path.basename(file_path).replace('.md', '')
            date = datetime.fromtimestamp(os.path.getmtime(file_path)).strftime('%Y-%m-%d')
        return {
            'title': title,
            'date': date,
        }

def generate_data_ts(md_files, filepath):
    data = [extract_metadata(file) for file in md_files]
    data.sort(key=lambda x: datetime.strptime(x['date'], '%Y-%m-%d'), reverse=True)
    
    ts_content = f"export const data = {json.dumps(data, indent=2)};"
    
    with open(filepath, 'w', encoding='utf-8') as f:
        f.write(ts_content)

if __name__ == '__main__':
    blogs_dir = 'pages/blogs'
    md_files = get_md_files(blogs_dir)
    generate_data_ts(md_files, 'data/blogs.ts')

    weekly_dir = 'pages/weekly'
    md_files = get_md_files(weekly_dir)
    generate_data_ts(md_files, 'data/weekly.ts')

    print("data.ts 已生成")
